### PR TITLE
Adding a restore workflow

### PR DESCRIPTION
**Pull request by MariamFahmy98**

### CODE REVIEW

File: src/include/node.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ * @param node The node node
+ */
+void pgmoneta_append_node(struct node** chain, struct node* node);
+
+/**
+ * Delete the node
+ * @param in The node
+ * @return 0 upon success, otherwise 1
+ */
+int pgmoneta_destroy_node(struct node* in);
```
Comment (jesperpedersen) [inline src/include/node.h:106]: `pgmoneta_free_nodes(`

File: src/libpgmoneta/backup.c [hunk 2]
```diff
@@ -59,6 +59,8 @@ pgmoneta_backup(int client_fd, int server, char** argv)
    unsigned long size;
    struct workflow* workflow = NULL;
    struct workflow* current = NULL;
+   struct node* i_nodes = NULL;
+   struct node** o_nodes = NULL;
```
Comment (jesperpedersen) [inline src/libpgmoneta/backup.c:63]: struct node* o_nodes = NULL;

File: src/libpgmoneta/backup.c [hunk 3]
```diff
@@ -96,7 +98,7 @@ pgmoneta_backup(int client_fd, int server, char** argv)
    current = workflow;
    while (current != NULL)
    {
-      if (current->setup(server, &date[0]))
+      if (current->setup(server, &date[0], i_nodes, o_nodes))
```
Comment (jesperpedersen) [inline src/libpgmoneta/backup.c:101]: `&o_nodes`

File: src/libpgmoneta/backup.c [hunk 4]
```diff
@@ -106,7 +108,7 @@ pgmoneta_backup(int client_fd, int server, char** argv)
    current = workflow;
    while (current != NULL)
    {
-      if (current->execute(server, &date[0]))
+      if (current->execute(server, &date[0], i_nodes, o_nodes))
```
Comment (jesperpedersen) [inline src/libpgmoneta/backup.c:111]: Ditto

File: src/libpgmoneta/backup.c [hunk 5]
```diff
@@ -116,7 +118,7 @@ pgmoneta_backup(int client_fd, int server, char** argv)
    current = workflow;
    while (current != NULL)
    {
-      if (current->teardown(server, &date[0]))
+      if (current->teardown(server, &date[0], i_nodes, o_nodes))
```
Comment (jesperpedersen) [inline src/libpgmoneta/backup.c:121]: Ditto

File: src/libpgmoneta/node.c [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+   memset(node, 0, sizeof(struct node));
+
+   node->data = malloc(strlen(s) + 1);
+   memset(node->data, 0, strlen(s) + 1);
+   memcpy(node->data, s, strlen(s));
+
+   node->tag = malloc(strlen(t) + 1);
+   memset(node->tag, 0, strlen(t) + 1);
+   memcpy(node->tag, t, strlen(t));
+
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:54]: `node->next = NULL;`

File: src/libpgmoneta/node.c [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+   memset(node, 0, sizeof(struct node));
+
+   node->data = malloc(sizeof(int));
+   memset(node->data, 0, sizeof(int));
+   memcpy(node->data, &val, sizeof(int));
+
+   node->tag = malloc(strlen(t) + 1);
+   memset(node->tag, 0, strlen(t) + 1);
+   memcpy(node->tag, t, strlen(t));
+
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:74]: Ditto

File: src/libpgmoneta/node.c [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+   memset(node, 0, sizeof(struct node));
+
+   node->data = malloc(sizeof(bool));
+   memset(node->data, 0, sizeof(bool));
+   memcpy(node->data, &val, sizeof(bool));
+
+   node->tag = malloc(strlen(t) + 1);
+   memset(node->tag, 0, strlen(t) + 1);
+   memcpy(node->tag, t, strlen(t));
+
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:94]: Ditto

File: src/libpgmoneta/node.c [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+   struct node* current = NULL;
+
+   current = chain;
+
+   while (current != NULL)
+   {
+
+      if (!strcmp(current->tag, t))
+      {
+         break;
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:131]: Ditto

File: src/libpgmoneta/node.c [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:152]: Ditto

File: src/libpgmoneta/node.c [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+      }
+
+      current = current->next;
+   }
+
+   return *(bool*)current->data;
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:158]: Lets return false here...

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
Review - part 1...

File: src/libpgmoneta/wf_gzip.c [hunk 12]
```diff
@@ -27,44 +27,57 @@
  */
 
 /* pgmoneta */
+#include "node.h"
```
Comment (jesperpedersen) [inline src/libpgmoneta/wf_gzip.c:30]:
```
#include "node.h" --> #include <node.h>
```

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/node.c [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+   if (node != NULL)
+   {
+      current = node;
+      nxt = current->next;
+
+      while (current != NULL)
+      {
+         nxt = current->next;
+
+         free(current);
```
Comment (jesperpedersen) [inline src/libpgmoneta/node.c:279]: You will need to free `data` and `tag` as well

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by MariamFahmy98** [comment]
> Review - part 1...

Do you mean the storage engine part ?

**Comment by MariamFahmy98** [comment]
The restore workflow doesn't work as expected. The restoration path sits empty after running the command.
I am still working on it.

**Comment by jesperpedersen** [comment]
No, just this pull request -- we need restore working again before we can consider merging

**Comment by MariamFahmy98** [comment]
I'm kind of lost in the code right now trying to figure out why the restore command isn't working. Can you spot where the problem is?

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
```
./pgmoneta-cli -c pgmoneta.conf restore primary latest /tmp
```

doesn't work. `position` is NULL which is valid.

Please, check all the various cases for `restore`

**Comment by jesperpedersen** [comment]
Ok.

However, the files are still compressed, so something isn't working in the chain

**Comment by jesperpedersen** [comment]
There are 2 `pgmoneta_delete_file()` in the decompress path...

**Pull request merged**